lua/ttt2/libraries/orm.lua

```diff
@@ -0,0 +1,219 @@
+---
+-- Object Relational Model
+-- @author Histalek
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+local sql = sql
+
+orm = orm or {}
+
+local ORMMODEL = {}
+
+---
+-- Returns an object relational model according to the specified databasetable. Does nothing if no databasetable with the given name exists.
+-- @param string tableName The name of the table in the database to create a model for.
+-- @param[opt] boolean force If set to `true` the function will not return a cached version of the model.
+-- @return ormmodel The model of the database table.
+-- @realm shared
+function orm.Make(tableName, force)
+	if IsValid(orm[tableName]) and not force then
+		return orm[tableName]
+	end
+
+	if not sql.TableExists(tableName) then return end
+
+	local model = {}
+	local primaryKey = sql.GetPrimaryKey(tableName)
+	local dataStructure = sql.GetTableColumns(tableName)
+
+	model._tableName = tableName
+	model._primaryKey = primaryKey
+	model._dataStructure = dataStructure
+	model.All = ORMMODEL.All
+	model.Find = ORMMODEL.Find
+	model.New = ORMMODEL.New
+
+	-- DO NOT setup delete/save functions if no primarykey is found.
+	-- In those cases the 'rowid' column would function as the primarykey, but as the rowid could change anytime (https://www.sqlite.org/rowidtable.html) data could be deleted unintentionally.
+	-- Most likely rowids wont change in gmod as there is no vacuum operation but just to be safe we will not allow to use such tables. ref: https://wiki.facepunch.com/gmod/sql
+
+	if not primaryKey then
+		return model
+	end
+
+	model.Delete = ORMMODEL.Delete
+	model.Save = ORMMODEL.Save
+	model.Refresh = ORMMODEL.Refresh
+
+	-- Prepare strings that will not change unless the model itself changes. So we don't have to create these strings everytime we use `model.Save()`.
+	local columnList = {nil, nil}
+
+	for i = 1, #primaryKey do
+		columnList[i] = sql.SQLIdent(primaryKey[i])
+	end
+
+	model._primaryKeyList = table.concat(columnList, ",")
+
+	for i = 1, #dataStructure do
+		columnList[i] = sql.SQLIdent(dataStructure[i])
+	end
+
+	model._columnList = table.concat(columnList, ",")
+
+	orm[tableName] = model
+
+	return model
+end
+
+---
+-- Retrieves all saved objects of the model from the database.
+-- @return table Returns an array of all found objects.
+-- @realm shared
+function ORMMODEL:All()
+	return sql.Query("SELECT * FROM " .. sql.SQLIdent(self._tableName))
+end
+
+---
+-- Deletes the given object from the database storage.
+-- @return nil|false Returns false if an error occurred, nil otherwise.
```

Comment:
i am only relaying the return value of the sql query. For a delete query this can only be either successful with no rows returned -> return nil or the sql query can fail for whatever reason -> false